source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,333 @@
+    parent_.flushPendingQueries(true /* with_error */);
+    // Note: Nothing can follow this call to flushPendingQueries due to deletion of this
+    // object upon resolution.
+    return;
+  }
+
+  // Only add this address to the list if kDNSServiceFlagsAdd is set. Callback targets are purely
+  // additive.
+  if (flags & kDNSServiceFlagsAdd) {
+    auto dns_response = buildDnsResponse(address, ttl);
+    ENVOY_LOG(debug, "Address to add address={}, ttl={}",
+              dns_response.address_->ip()->addressAsString(), ttl);
+
+    if (!pending_cb_) {
+      ENVOY_LOG(debug, "Adding to queries pending callback");
+      pending_cb_ = {ResolutionStatus::Success, {dns_response}};
+      parent_.addPendingQuery(this);
+    } else {
+      ENVOY_LOG(debug, "New address for query already pending flush");
+      pending_cb_->responses_.push_back(dns_response);
+    }
+  }
+
+  if (!(flags & kDNSServiceFlagsMoreComing)) {
+    /* (taken and edited from dns_sd.h)
+     * Collective kDNSServiceFlagsMoreComing flag:
+     * When [DNSServiceGetAddrInfoReply] are invoked using a shared DNSServiceRef, the
+     * kDNSServiceFlagsMoreComing flag applies collectively to *all* active
+     * operations sharing the same [main_sd_ref]. If the MoreComing flag is
+     * set it means that there are more results queued on this parent DNSServiceRef,
+     * but not necessarily more results for this particular callback function.
+     * The implication of this for client programmers is that when a callback
+     * is invoked with the MoreComing flag set, the code should update its
+     * internal data structures with the new result (as is done above when calling
+     * parent_.addPendingQuery(this))...Then, later when a callback is eventually invoked with the
+     * MoreComing flag not set, the code should update *all* [pending queries] related to that
+     * shared parent DNSServiceRef that need updating (i.e that have had DNSServiceGetAddrInfoReply
+     * called on them since the last flush), not just the [queries] related to the particular
+     * callback that happened to be the last one to be invoked.
+     */
+    ENVOY_LOG(debug, "DNS Resolver flushing queries pending callback");
+    parent_.flushPendingQueries(false /* with_error */);
+    // Note: Nothing can follow this call to flushPendingQueries due to deletion of this
+    // object upon resolution.
+    return;
+  }
+}
+
+DNSServiceErrorType
+AppleDnsResolverImpl::PendingResolution::dnsServiceGetAddrInfo(DnsLookupFamily dns_lookup_family) {
+  DNSServiceProtocol protocol;
+  switch (dns_lookup_family) {
+  case DnsLookupFamily::V4Only:
+    protocol = kDNSServiceProtocol_IPv4;
+    break;
+  case DnsLookupFamily::V6Only:
+    protocol = kDNSServiceProtocol_IPv6;
+    break;
+  case DnsLookupFamily::Auto:
+    protocol = kDNSServiceProtocol_IPv4 | kDNSServiceProtocol_IPv6;
+    break;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
```

Comment:
Is this needed? Don't you cover all cases?

Reply:
I generally leave this defensively for enums even if I have exhausted at present in case people add more values. But I guess this is an enum that would not be expanded lightly 😂